src/lib/components/Swap/SwapButton.tsx

```diff
@@ -61,72 +59,73 @@ export default function SwapButton({ disabled }: SwapButtonProps) {
     useSwapApprovalOptimizedTrade(trade.trade, allowedSlippage, useIsPendingApproval) || trade.trade
 
   const approvalCurrencyAmount = useSwapCurrencyAmount(Field.INPUT)
-  const [approval, getApproval] = useSwapApproval(
-    optimizedTrade,
-    allowedSlippage,
-    useIsPendingApproval,
-    approvalCurrencyAmount
-  )
+
+  const { notApproved, pendingApproval, loadingSignature, supportsPermit, signatureData, handleApproveOrPermit } =
+    useApproveAndPermit(optimizedTrade, allowedSlippage, useIsPendingApproval, approvalCurrencyAmount)
+
   const approvalHash = usePendingApproval(
     inputCurrency?.isToken ? inputCurrency : undefined,
     useSwapRouterAddress(optimizedTrade)
   )
 
   const addTransaction = useAddTransaction()
-  const addApprovalTransaction = useCallback(() => {
-    getApproval().then((transaction) => {
-      if (transaction) {
-        addTransaction({ type: TransactionType.APPROVAL, ...transaction })
-      }
-    })
-  }, [addTransaction, getApproval])
+  const deadline = useTransactionDeadline()
 
   const actionProps = useMemo((): Partial<ActionButtonProps> | undefined => {
-    if (!disabled && chainId) {
-      if (approval === ApprovalState.NOT_APPROVED) {
-        const currency = inputCurrency || approvalCurrencyAmount?.currency
-        invariant(currency)
-        return {
-          action: {
-            message: <Trans>Approve {currency.symbol} first</Trans>,
-            onClick: addApprovalTransaction,
-            children: <Trans>Approve</Trans>,
-          },
-        }
-      } else if (approval === ApprovalState.PENDING) {
-        return {
-          disabled: true,
-          action: {
-            message: (
-              <EtherscanLink type={ExplorerDataType.TRANSACTION} data={approvalHash}>
-                <Trans>Approval pending</Trans>
-              </EtherscanLink>
-            ),
-            icon: Spinner,
-            onClick: addApprovalTransaction,
-            children: <Trans>Approve</Trans>,
-          },
-        }
-      } else if (inputCurrencyAmount && inputCurrencyBalance && !inputCurrencyBalance.lessThan(inputCurrencyAmount)) {
-        return {}
+    if (
+      disabled ||
+      !chainId ||
+      loadingSignature ||
+      (inputCurrencyAmount && inputCurrencyBalance && inputCurrencyBalance.lessThan(inputCurrencyAmount))
+    ) {
+      return { disabled: true }
+    }
+    if (notApproved) {
+      const currency = inputCurrency || approvalCurrencyAmount?.currency
+      invariant(currency)
```

Comment:
Will using invariant here throw to the worst case error screen?

Reply:
i think so? this logic was just taken from existing approval logic - so maybe thats a question for another PR?

Reply:
yes it will. I believe that I put that here. The other option is to check that `currency` is defined, but I think that with the logic above that's already checked for. It may be possible to rewrite these conditionals so that the typechecker understands all of this betetr.